File: excelium-cli/src/main/java/excelium/cli/BeanFactory.java
/*
 * MIT License
 *
 * Copyright (c) 2018 Excelium
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package excelium.cli;

import com.google.api.services.sheets.v4.Sheets;
import excelium.cli.annotation.Bean;
import excelium.common.XmlMarshaller;
import excelium.core.reader.TestReaderFactory;
import excelium.core.writer.TestWriterFactory;
import excelium.model.enums.WorkbookType;
import excelium.model.project.Project;
import excelium.sheets.SheetsReaderFactory;
import excelium.sheets.SheetsServiceProvider;
import excelium.sheets.SheetsWriterFactory;
import excelium.sheets.connection.GoogleConnection;
import excelium.sheets.connection.GoogleConnectionService;
import excelium.xls.ExcelReaderFactory;
import excelium.xls.ExcelWriterFactory;

import javax.xml.bind.JAXBException;
import java.io.IOException;
import java.nio.file.Paths;

/**
 * Provides bean factory methods for constructing injectable application beans.
 *
 * @author PhungDucKien
 * @since 2018.04.10
 */
public class BeanFactory {

    /**
     * Reads the current working directory's project configuration file and constructs project instance.
     *
     * @return Project instance
     * @throws JAXBException the jaxb exception
     */
    @Bean
    public Project getProject() throws JAXBException {
        return (Project) XmlMarshaller.unmarshall(Paths.get("project.xml"), Project.class);
    }

    /**
     * Constructs a Google connection.
     *
     * @param project Project instance
     * @return Google connection if project uses Sheets file. Otherwise, return null.
     */
    @Bean
    public GoogleConnection getGoogleConnection(Project project) {
        if (project.getWorkbookType() == WorkbookType.SHEETS) {
            return new GoogleConnectionService();
        }
        return null;
    }

    /**
     * Constructs Sheets service.
     *
     * @param project    Project instance
     * @param connection Google connection
     * @return Sheets service if project uses Sheets file. Otherwise, return null.
     * @throws IOException if IOException occurs
     */
    @Bean
    public Sheets getSheetsService(Project project, GoogleConnection connection) throws IOException {
        if (project.getWorkbookType() == WorkbookType.SHEETS && connection != null) {
            SheetsServiceProvider sheetsServiceProvider = new SheetsServiceProvider(connection);
            return sheetsServiceProvider.createSheetsService();
        }
        return null;
    }

    /**
     * Constructs test reader factory
     *
     * @param project       Project instance
     * @param sheetsService Sheets service
     * @return Excel test reader factory if project uses Excel files. Sheets test reader factory if projects uses Sheets files.
     */
    @Bean
    public TestReaderFactory getTestReaderFactory(Project project, Sheets sheetsService) {
        if (project.getWorkbookType() == WorkbookType.SHEETS) {
            return new SheetsReaderFactory(sheetsService);
        } else {
            return new ExcelReaderFactory();
        }
    }

    /**
     * Constructs test writer factory
     *
     * @param project       Project instance
     * @param sheetsService Sheets service
     * @return Excel test writer factory if project uses Excel files. Sheets test writer factory if projects uses Sheets files.
     */
    @Bean
    public TestWriterFactory getTestWriterFactory(Project project, Sheets sheetsService) {
        if (project.getWorkbookType() == WorkbookType.SHEETS) {
            return new SheetsWriterFactory(sheetsService);
        } else {
            return new ExcelWriterFactory();
        }
    }
}

File: packages/cli/src/BeanFactory.ts
/*
 * Licensed to the Software Freedom Conservancy (SFC) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The SFC licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { JsonMarshaller, StringUtil } from '@excelium/common';
import { GenericTestReaderFactory, GenericTestWriterFactory } from '@excelium/core';
import { DataSource, Project, Template, TestFile, WorkbookType } from '@excelium/model';
import { GoogleConnection, GoogleConnectionService, SheetsReaderFactory, SheetsServiceProvider, SheetsWriterFactory } from '@excelium/sheets';
import { ExcelReaderFactory, ExcelWriterFactory } from '@excelium/xls';
import { plainToClass } from 'class-transformer';
import { sheets_v4 } from 'googleapis';
import { interfaces } from 'inversify';
import path from 'path';

/**
 * Provides bean factory methods for constructing injectable application beans.
 *
 * @author PhungDucKien
 * @since 2018.04.10
 */
export default class BeanFactory {
  /**
   * Reads the current working directory's project configuration file and constructs project instance.
   *
   * @return Project instance
   */
  public static projectProvider = (customProjectFile?: string) => {
    return () => {
      return async (): Promise<Project> => {
        const projectFilePath = !customProjectFile || StringUtil.isBlank(customProjectFile) ? 'excelium.json' : customProjectFile;
        const project = plainToClass(Project, await JsonMarshaller.unmarshall(projectFilePath), {
          targetMaps: [
            {
              target: Project,
              properties: {
                tests: TestFile,
                templates: Template,
                dataSources: DataSource,
              },
            },
          ],
        });
        project.basePath = path.dirname(path.resolve(projectFilePath));
        return project;
      };
    };
  };

  /**
   * Constructs a Google connection.
   *
   * @param project Project instance
   * @return Google connection if project uses Sheets file. Otherwise, return null.
   */
  public static googleConnectionProvider = () => {
    return (context: interfaces.Context) => {
      return async (): Promise<GoogleConnection | null> => {
        const projectProvider = context.container.get<() => Promise<Project>>(BeanFactory.projectProvider);
        const project = await projectProvider();
        if (project.workbookType === WorkbookType.SHEETS) {
          return new GoogleConnectionService();
        }
        return null;
      };
    };
  };

  /**
   * Constructs Sheets service.
   *
   * @param project    Project instance
   * @param connection Google connection
   * @return Sheets service if project uses Sheets file. Otherwise, return null.
   */
  public static sheetsServiceProvider = () => {
    return (context: interfaces.Context) => {
      return async (): Promise<sheets_v4.Sheets | null> => {
        const projectProvider = context.container.get<() => Promise<Project>>(BeanFactory.projectProvider);
        const project = await projectProvider();
        const googleConnectionProvider = context.container.get<() => Promise<GoogleConnection | null>>(BeanFactory.googleConnectionProvider);
        const googleConnection = await googleConnectionProvider();

        if (project.workbookType === WorkbookType.SHEETS && googleConnection != null) {
          const sheetsServiceProvider = new SheetsServiceProvider(googleConnection);
          return sheetsServiceProvider.createSheetsService();
        }
        return null;
      };
    };
  };

  /**
   * Constructs test reader factory
   *
   * @param project       Project instance
   * @param sheetsService Sheets service
   * @return Excel test reader factory if project uses Excel files. Sheets test reader factory if projects uses Sheets files.
   */
  public static testReaderFactoryProvider = () => {
    return (context: interfaces.Context) => {
      return async (): Promise<GenericTestReaderFactory> => {
        const projectProvider = context.container.get<() => Promise<Project>>(BeanFactory.projectProvider);
        const project = await projectProvider();
        const sheetsServiceProvider = context.container.get<() => Promise<sheets_v4.Sheets | null>>(BeanFactory.sheetsServiceProvider);
        const sheetsService = await sheetsServiceProvider();

        if (project.workbookType === WorkbookType.SHEETS && sheetsService != null) {
          return new SheetsReaderFactory(sheetsService);
        } else {
          return new ExcelReaderFactory();
        }
      };
    };
  };

  /**
   * Constructs test writer factory
   *
   * @param project       Project instance
   * @param sheetsService Sheets service
   * @return Excel test writer factory if project uses Excel files. Sheets test writer factory if projects uses Sheets files.
   */
  public static testWriterFactoryProvider = () => {
    return (context: interfaces.Context) => {
      return async (): Promise<GenericTestWriterFactory> => {
        const projectProvider = context.container.get<() => Promise<Project>>(BeanFactory.projectProvider);
        const project = await projectProvider();
        const sheetsServiceProvider = context.container.get<() => Promise<sheets_v4.Sheets | null>>(BeanFactory.sheetsServiceProvider);
        const sheetsService = await sheetsServiceProvider();

        if (project.workbookType === WorkbookType.SHEETS && sheetsService != null) {
          return new SheetsWriterFactory(sheetsService);
        } else {
          return new ExcelWriterFactory();
        }
      };
    };
  };
}
